h2o-core/src/main/java/water/jdbc/SQLManager.java

```diff
@@ -245,7 +247,9 @@ public SqlTableToH2OFrame(String url, String table, boolean needFetchClause, Str
     @Override
     protected void setupLocal() {
       int conPerNode = (int) Math.min(Math.ceil((double) _nChunks / H2O.getCloudSize()), Runtime.getRuntime().availableProcessors());
-      conPerNode = Math.min(conPerNode, SQLManager.MAX_CONNECTIONS / H2O.getCloudSize()); 
+      conPerNode = Math.min(conPerNode, SQLManager.MAX_CONNECTIONS / H2O.getCloudSize());
+      //Make sure at least some connections are available to a node
+      conPerNode = Math.max(conPerNode, MIN_CONNECTIONS_PER_NODE);
```

Comment:
Make sure there is at least MIN_CONNECTIONS_PER_NODE number of connections. Currently, the value is set to 1. This may produce connection "pools" larger than 100. Each node must have at least 1 connection available for us to work correctly.

- A database backing such large cluster should be dimensioned accordingly.
- Original formula for "normal-sized" clusters does not change.

For large clusters, we could use different strategies, like synchronized opening/closing of connections not to exceed the limit, however there is overhead to that as well.

Reply:
Makes sense!